Add a new page with guidance on how to write UI/e2e tests for a Teams application

adinastoll opened the pull request:
# Pull Request Template

## What are you trying to address
Contributing with a new guidance based on previous engagement on how to write UI tests for a custom Teams application.

## Checklist

- [x] Changes follow the repo structure and land in the appropriate folder and section
- [x] No confidential information
- [x] No duplicated content
- [x] Labeled appropriately
- [x] Added 2 reviewers
- [x] No lint errors
- [x] No lint check errors related to your changes
      
> Note: You may see link check errors on pages you have not touched. This is normal, and due to either broken links or sites that reject link checker bots. The reviewer will help you get to a green state on these.

benjguin commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,230 @@
+# UI Tests for a Teams application
```
```suggestion
# Automated UI Tests for a Teams application
```

benjguin reviewed (approved):
The content looks good to me.

A few remarks:

1/ Maybe the step by step instructions are too detailed for the playbook and may generate significant maintenance work of this page over time.
I'm not managing the playbook content, so I'm not sure.

2/ Maybe a chapter or a paragraph in the Overview should summarize the findings. I think of:
- testing in Web Browser is simpler than on a native app
- Testing in a local Windows box is more stable than in a CICD that involves a CICD agent plus a VM in VM for the emulator
- Testing in a Teams app in a native app is hard
  - you are testing an app in Teams App, not the Teams App for which you don't have the APK. Frameworks expoct you to test the app for which you have the APK
  - it requires the installation of the app which is hard to automate because it requires a Google account to access the Play store

Things of that level

adinastoll commented on docs/automated-testing/ui-testing/teams-tests.md:
````diff
@@ -0,0 +1,230 @@
+```
+If you would like to test the production version of your app which was already published in the store you would first need to automate the installation of the app from Teams' store for your test user.
+
+Once the app was installed, implement selectors to [access your custom app](https://learn.microsoft.com/en-us/microsoftteams/platform/concepts/deploy-and-publish/apps-upload#access-your-app) and to perform various actions within the app. 
````
That's right. I added the provided link for reference. Thank you!

adinastoll commented:
@benjguin Thank you for the review! :) Based on your suggestion, I added a sub-chapter in Overview to highlight the learnings. Thanks for the suggestion! 

I'll keep the step by steps until further feedback from Playbook maintainers, though your point is really valid. I might keep just the aspects which are relevant to Teams (such as choosing a device with Play Store enabled, else Teams can't be installed etc)

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+
+<!-- There are two proven ways through which this can still be achieved for Android, based on previous experience with our customers: -->
+
+### 1. Using Android Virtual Devices
```
```suggestion
### Using Android Virtual Devices
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+        > [Here](https://appium.io/docs/en/2.0/ecosystem/) you can read more about Appium ecosystem
+      - The advantage of this architecture is that it opens the possibility of running the server in a VM, and the client in a pipeline, enabling the tests to be ran automatically on scheduled basis as part of CI/CD pipelines. 
+
+### How to run mobile test locally on a Windows machine 
+In order to run locally such tests, you need two things:
+1. An emulator ([AVD - Android Virtual Devices](https://developer.android.com/studio/run/managing-avds)), which will represent the physical device.
+1. [Appium server](https://appium.io/docs/en/2.1/), which will redirect the commands from the test to your virtual device.
```
```suggestion
### Running mobile test locally on a Windows machine 

This requires:

- An emulator ([AVD - Android Virtual Devices](https://developer.android.com/studio/run/managing-avds)), which will represent the physical device.
- [Appium server](https://appium.io/docs/en/2.1/), which will redirect the commands from the test to your virtual device.
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+1. An emulator ([AVD - Android Virtual Devices](https://developer.android.com/studio/run/managing-avds)), which will represent the physical device.
+1. [Appium server](https://appium.io/docs/en/2.1/), which will redirect the commands from the test to your virtual device.
+
+#### a) Steps to create Android Virtual Device:
+1. Install Android Studio from [official link](https://developer.android.com/studio). 
+    > Note: At the time of writing the documentation, the latest version available was Android Studio Giraffe, 2022.3.1 Patch 2 for Window.
+    1. Set ANDROID_HOME environment variable to point to the installation path of Android SDK. i.e. ` C:Users\<user-name>\AppData\Local\Android\Sdk`
+1. Install Java Development Kit (JDK) from [official link](https://www.oracle.com/java/technologies/javase/jdk11-archive-downloads.html). For the most recent devices JDK 9 is required, otherwise JDK 8 is required. Make sure you get the JDK and not the JRE.
+    1. Set JAVA_HOME environment variable to the installation path, i.e. ` C:\Program Files\Java\jdk-11`
+1. Create an AVD (Android Virtual Device):
+    1. Open Android Studio. From the Android Studio welcome screen, select **More Action -> Virtual Device Manager**, as instructed [here](https://developer.android.com/studio/run/managing-avds#:~:text=1%20Open%20the%20AVD%20Manager%20by%20clicking%20Tools,%20settings%2C%20such%20as%20the%20skin.%20See%20More.)
+    1. Click **Create Device**.
+    1. Choose a device definition with **Play Store** enabled. This is important, otherwise Teams cannot be installed on the device.
+    1. Choose a System image from the **Recommended** tab which includes access to Google Play services. You may need to install it before selecting it.
+    1. Start the emulator by clicking on the Run button from the Device Manage screen.
+    1. Manually install Microsoft Teams from Google Playstore on the device. 
```
```suggestion
#### Creating an Android Virtual Device:

1. Install Android Studio from [official link](https://developer.android.com/studio). 
    > Note: At the time of writing the documentation, the latest version available was Android Studio Giraffe, 2022.3.1 Patch 2 for Window.
    1. Set ANDROID_HOME environment variable to point to the installation path of Android SDK. i.e. ` C:Users\<user-name>\AppData\Local\Android\Sdk`
1. Install Java Development Kit (JDK) from [official link](https://www.oracle.com/java/technologies/javase/jdk11-archive-downloads.html). For the most recent devices JDK 9 is required, otherwise JDK 8 is required. Make sure you get the JDK and not the JRE.
    1. Set JAVA_HOME environment variable to the installation path, i.e. ` C:\Program Files\Java\jdk-11`
1. Create an AVD (Android Virtual Device):
    1. Open Android Studio. From the Android Studio welcome screen, select **More Action -> Virtual Device Manager**, as instructed [here](https://developer.android.com/studio/run/managing-avds#:~:text=1%20Open%20the%20AVD%20Manager%20by%20clicking%20Tools,%20settings%2C%20such%20as%20the%20skin.%20See%20More.)
    1. Click **Create Device**.
    1. Choose a device definition with **Play Store** enabled. This is important, otherwise Teams cannot be installed on the device.
    1. Choose a System image from the **Recommended** tab which includes access to Google Play services. You may need to install it before selecting it.
    1. Start the emulator by clicking on the Run button from the Device Manage screen.
    1. Manually install Microsoft Teams from Google Playstore on the device. 
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+    1. Start the emulator by clicking on the Run button from the Device Manage screen.
+    1. Manually install Microsoft Teams from Google Playstore on the device. 
+
+####  b) Steps to set up Appium
+- Install `appium`:
+    1. Download NodeJs, if it is not already installed on your machine: [Download | Node.js (nodejs.org)](https://nodejs.org/en/download)
+    1. Install Appium globally: [Install Appium - Appium Documentation](https://appium.io/docs/en/2.0/quickstart/install/)
+    1. Install the UiAutomator2 driver: [Install the UiAutomator2 Driver - Appium Documentation](https://appium.io/docs/en/2.0/quickstart/uiauto2-driver/). Go through the `Set up Android automation requirements` in the documentation, to make sure you have set up everything correctly.
+
+        > [Here](https://appium.io/docs/en/2.0/intro/drivers/) you can read more about Appium Drivers.
+    1. Start appium server by running `appium` command in a command prompt.
```
```suggestion
####  Setting up Appium
- Install `appium`:
    1. Download NodeJs, if it is not already installed on your machine: [Download | Node.js (nodejs.org)](https://nodejs.org/en/download)
    1. Install Appium globally: [Install Appium - Appium Documentation](https://appium.io/docs/en/2.0/quickstart/install/)
    1. Install the UiAutomator2 driver: [Install the UiAutomator2 Driver - Appium Documentation](https://appium.io/docs/en/2.0/quickstart/uiauto2-driver/). Go through the `Set up Android automation requirements` in the documentation, to make sure you have set up everything correctly.

        > [Here](https://appium.io/docs/en/2.0/intro/drivers/) you can read more about Appium Drivers.
    1. Start appium server by running `appium` command in a command prompt.
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
````diff
@@ -0,0 +1,240 @@
+
+        > [Here](https://appium.io/docs/en/2.0/intro/drivers/) you can read more about Appium Drivers.
+    1. Start appium server by running `appium` command in a command prompt.
+####  Useful commands:
+- To list the emulators that you have previously created, without opening Android Studio:
+    ```cli
+    emulator -list-avds
+    ```
+- To start an emulator, without opening Android Studio:
+    ```cli
+    emulator -avd {name-of-emulator}
+    ```
+- To get the name of your AVD, assuming your emulator is running, you can run:
+    
+    ```cli
+    adb devices
+    ```
+    You should see a response with *List of devices attached*, where your running device will be listed.
````
```suggestion
####  Useful commands:

List emulators that you have previously created, without opening Android Studio:

```cli
emulator -list-avds
```

Start an emulator, without opening Android Studio:
    
```cli
emulator -avd {name-of-emulator}
```

Get the name of your AVD, assuming your emulator is running, you can run:
    
```cli
adb devices
```

You should see the response *List of devices attached*, where your running device will be listed.
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
````diff
@@ -0,0 +1,240 @@
+    ```
+    You should see a response with *List of devices attached*, where your running device will be listed.
+
+### How to set up an Azure VM for running mobile tests in a pipeline
````
```suggestion
### Setting up an Azure VM for running mobile tests in a pipeline
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+
+### How to set up an Azure VM for running mobile tests in a pipeline
+
+### 1. VM configuration
+This approach essentially hosts a virtual device in a virtual machine. In order to be able to set up the emulator (Android Virtual Device) in an Azure VM, the VM should support [nested virtualization](https://azure.microsoft.com/en-us/blog/nested-virtualization-in-azure/). 
```
```suggestion
### Configure the VM

This approach essentially hosts a virtual device in a virtual machine. In order to be able to set up the emulator (Android Virtual Device) in an Azure VM, the VM should support [nested virtualization](https://azure.microsoft.com/en-us/blog/nested-virtualization-in-azure/). 
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+- VM generation: V1
+- Size: Standard D4ds v5 16 GiB memory
+
+### 2. Enable connection from outside to Appium server on the VM 
+> Note: By default appium server runs on port 4723. The rest of the steps will assume that this is the port where your appium server runs.
```
```suggestion
### Enable connection from outside to Appium server on the VM 

> Note: By default appium server runs on port 4723. The rest of the steps will assume that this is the port where your appium server runs.
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+### 2. Enable connection from outside to Appium server on the VM 
+> Note: By default appium server runs on port 4723. The rest of the steps will assume that this is the port where your appium server runs.
+
+In order to be able to reach appium server which runs on the VM from outside, the following steps are needed:
+1. Create an [Inbound Rule](https://learn.microsoft.com/en-us/windows/security/operating-system-security/network-security/windows-firewall/create-an-inbound-port-rule) for port 4723 from within the VM.
```
```suggestion
In order to be able to reach appium server which runs on the VM from outside:

1. Create an [Inbound Rule](https://learn.microsoft.com/en-us/windows/security/operating-system-security/network-security/windows-firewall/create-an-inbound-port-rule) for port 4723 from within the VM.
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+    - Find out the IP of the machine on which the tests will run on.
+    - Replace the *Source IP Address* with the IP of your machine.
+
+### 3. Install Android Studio and create AVD inside the VM
+1. Follow the instructions from steps a) and b) in **How to run mobile end to end tests on a Windows machine** section above.
```
```suggestion
### Installing Android Studio and create AVD inside the VM

1. Follow the instructions from steps a) and b) in **How to run mobile end to end tests on a Windows machine** section above.
```
```suggestion
### 3. Install Android Studio and create AVD inside the VM
1. Follow the instructions from steps a) and b) in **How to run mobile end to end tests on a Windows machine** section above.
```
```suggestion
### Install Android Studio and create AVD inside the VM

1. Follow the instructions from steps a) and b) in **How to run mobile end to end tests on a Windows machine** section above.
```

TessFerrandez replied:
NOTE: This section does not exist above... Point to the correct section and link with the #link

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+1. Restart the AVD.
+
+
+### How to inspect the Teams app in Azure Virtual Device (AVD)
+Inspecting the app is very useful as you write new tests, as it allows you to find the ID of different elements displayed on the screen - similar to DevTools, which allows you to navigate through the DOM of a web page.
```
```suggestion
### Inspecting the Teams app in Azure Virtual Device (AVD)

Inspecting the app is very useful as you write new tests, as it allows you to find the ID of different elements displayed on the screen - similar to DevTools, which allows you to navigate through the DOM of a web page.
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+
+> Note: This section assumes that you have already performed the prerequisites from **How to run mobile end to end tests on a Windows machine**
+
+### Steps to inspect Teams app:
+1. Run the appium server with [--alow-cors flag](https://appium.readthedocs.io/en/latest/en/writing-running-appium/server-args/) by running the following command in a terminal:
```
```suggestion
### Inspect the Teams app:

1. Run the appium server with [--alow-cors flag](https://appium.readthedocs.io/en/latest/en/writing-running-appium/server-args/) by running the following command in a terminal:
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+    ![teams-appium-inspector](./images/teams-appium-inspector.png)
+    - You can do any action on the emulator, and if you press on the "Refresh" button in the browser, the left hand side of the Appium Inspector will reflect your app. In the **App Source** you will be able to see the IDs of the elements, so you can write relevant selectors in your tests. 
+
+#### How to connect to Appium server
+We will provide an example on how to achieve this using JavaScript. Similar approach can be followed for other languages.
```
```suggestion
#### Connecting to Appium server

Below is how this can be accomplished with JavaScript. A similar approach can be followed for other languages.
```

TessFerrandez commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,240 @@
+1. "appium:appActivity": the activity within Teams that you would like to launch on the device. In our case, we would like just to launch the app. The activity name for launching Teams is called "com.microsoft.skype.teams.Launcher".
+1. "appium:automationName": the name of the driver you are using. Note: Appium can communicate to different platforms. This is achieved by installing a dedicated driver, designed for each platform. In our case, it would be [UiAutomator2](https://github.com/appium/appium-uiautomator2-driver) or [Espresso](https://github.com/appium/appium-espresso-driver), since they are both designed for Android p ...
+
+### 2. Using BrowserStack
+- BrowserStack does not support out of the box the installation of Teams from the App Store or Play Store. However, there is a workaround, described in [their documentation](https://www.browserstack.com/support/faq/app-automate/app/can-i-install-an-app-from-the-app-store-or-play-store). Therefore, if you choose to go this way, you would first need to implement a step that installs Teams on the cl ...
```
```suggestion
### Using BrowserStack

- BrowserStack does not support out of the box the installation of Teams from the App Store or Play Store. However, there is a workaround, described in [their documentation](https://www.browserstack.com/support/faq/app-automate/app/can-i-install-an-app-from-the-app-store-or-play-store). Therefore, if you choose to go this way, you would first need to implement a step that installs Teams on the cloud-based device, by implementing the workaround described above.
```

TessFerrandez reviewed (commented):
Great content... I have added some language suggestions and general formatting

adinastoll commented:
> Great content... I have added some language suggestions and general formatting
 
Thank you @TessFerrandez for your review! I have addressed the comments. :)

TessFerrandez commented:
@shiranr or @tompaana could one of you review this?

shiranr commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,235 @@
+# Automated UI Tests for a Teams application
```
Is the images self generated?

adinastoll replied:
yeah, they are self generated

shiranr commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,235 @@
+
+## Web based UI tests 
+
+To implement web-based UI tests for your Teams application, you can follow the same approach as for testing any other web application with a UI. [UI testing](./README.md) provides good guidance on that. The starting point for your test would be to launch a browser in an automated way (using Selenium, or similar frameworks) and go to [https://teams.microsoft.com/](https://teams.microsoft.com/). 
```
```suggestion
To implement web-based UI tests for your Teams application, you can follow the same approach as for testing any other web application with a UI. [UI testing](README.md) provides good guidance on that. The starting point for your test would be to launch a browser in an automated way (using Selenium, or similar frameworks) and go to [https://teams.microsoft.com/](https://teams.microsoft.com/). 
```

shiranr commented on docs/automated-testing/ui-testing/teams-tests.md:
````diff
@@ -0,0 +1,235 @@
+If you would like to test a Teams app which was not yet published into the Teams' store, or if you would like to test the DEV/QA version of your app:
+- Using for example [Teams Toolkit](https://github.com/OfficeDev/TeamsFx), package your app based on the [manifest.json](https://learn.microsoft.com/en-us/microsoftteams/platform/resources/schema/manifest-schema). See below an example on how to achieve that:
+
+```javascript
+npx teamsfx package --env dev --manifest-path ...
+
````
```suggestion
```javascript
npx teamsfx package --env dev --manifest-path ...
```
```

shiranr replied:
I am not sure it got my suggestion to add ``` at the end of the block as it currently doesn't exist.

adinastoll replied:
Thanks! Resolved

shiranr commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,235 @@
+# Automated UI Tests for a Teams application
```
Please notice these added words are causing the linter to break:
Unknown word (teamsfx)    

Unknown word (XVFB)       

Unknown word (framebuffer) 

Unknown word (Xvfb)        

Unknown word (XVFB)        

Unknown word (Xvfb)        

Unknown word (inprivate)   

Unknown word (Automator)   

Unknown word (Playstore)   

Unknown word (Automator)   

Unknown word (Automator)   

Unknown word (avds)        

Unknown word (runing)      

Unknown word (Automator)   

Unknown word (Automator)   

Unknown word (printscren)  

Unknown word (webdriverio) 

Unknown word (wdio)

Can you please add those to the .cspell?

shiranr commented on docs/automated-testing/ui-testing/teams-tests.md:
````diff
@@ -0,0 +1,235 @@
+
+List emulators that you have previously created, without opening Android Studio:
+
+```cli
+emulator -list-avds
````
```suggestion
```cli
emulator -list-avds
```
```

shiranr replied:
This is missing a closing ```

adinastoll replied:
Thanks! Resolved

shiranr commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,235 @@
+1. Follow the instructions from steps a) and b) in **How to run mobile end to end tests on a Windows machine** section above.
+1. When you launch the emulator, it may show a warning as below and will eventually crash:
+
+    ![failure](./images/warning.png)
```
```suggestion
    ![failure](images/warning.png)
```

adinastoll replied:
Thanks! Resolved

shiranr commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,235 @@
+    If the appium server runs on your local machine at the default portal, then Remote Host and Remote Port can be kept to the default values.
+
+    The configuration should look similar to the printscren below:
+    ![appium-inspector](./images/appium-inspector.png)
```
```suggestion
    ![appium-inspector](images/appium-inspector.png)
```

adinastoll replied:
Thanks! Resolved

shiranr commented on docs/automated-testing/ui-testing/teams-tests.md:
```diff
@@ -0,0 +1,235 @@
+    ![appium-inspector](./images/appium-inspector.png)
+3. Press on **Start Session**.
+    - In the browser, you should see a similar view as below:
+    ![teams-appium-inspector](./images/teams-appium-inspector.png)
```
```suggestion
    ![teams-appium-inspector](images/teams-appium-inspector.png)
```

adinastoll replied:
Thanks! Resolved

shiranr commented on docs/automated-testing/ui-testing/teams-tests.md:
````diff
@@ -0,0 +1,235 @@
+    await wdio.remote(opts);
+```
+
+1. "port": the port on which the Appium server runs on. By default, it is 4723.
+1. "hostname": the IP of the machine where the Appium sever runs on. If it is running locally, that is 127.0.0.1. If it runs in an Azure VM, it would be the public IP address of the VM. Note: ensure you have followed the steps from **2. Enable connection from outside to Appium server on the VM**.
+1. "platformName": Appium can be used to connect to different platforms (Windows, iOS, Android). In our case, it would be "android".
+1. "appium:deviceName": the name of the Android Virtual Device. See **Useful commands** on how to find the name of the device.
+1. "appium:appPackage": the name of the app's package that you would like to launch. Teams' package name is "com.microsoft.teams".
+1. "appium:appActivity": the activity within Teams that you would like to launch on the device. In our case, we would like just to launch the app. The activity name for launching Teams is called "com.microsoft.skype.teams.Launcher".
+1. "appium:automationName": the name of the driver you are using. Note: Appium can communicate to different platforms. This is achieved by installing a dedicated driver, designed for each platform. In our case, it would be [UiAutomator2](https://github.com/appium/appium-uiautomator2-driver) or [Espresso](https://github.com/appium/appium-espresso-driver), since they are both designed for Android p ...
````
Is this text taken from somewhere or is it self originated?

adinastoll replied:
@shirar thanks for reviewing it! It is self originated

shiranr commented:
Great content! thank you for contributing :)
Left couple of comments.

adinastoll commented:
> Great content! thank you for contributing :) Left couple of comments.

Thank you for reviewing it, @shiranr. I have addressed all your comments! Please have a look.

shiranr reviewed (approved):
LGTM! thanks
